### PR TITLE
Handling database auto re-connection

### PR DESCRIPTION
**Pull request by davidlatwe**
This will handling MongoDB auto re-connection automatically ( 3 retry times ), especially useful if you work with MongoDB Replica Set ( Auto re-connect if the primary node went down ).

( I have changed to use Replica Set recently. )

### CODE REVIEW

**Comment by BigRoy** [comment]
When exactly do you need this? When have you been facing the issue of disconnection? Somehow I've never seen an issue like that and I am wondering why this is a required workaround. Could you elaborate more on the case(s) where this is helpful?

**Comment by BigRoy** [comment]
Ah, that makes sense. Though trying three reconnections in this brief time, is that sufficient? Taking a quick glimpse at Replica Sets in Mongo it seemst that the average election for a new primary MongoDB instance being completed takes around 10 seconds?

And to be sure, this change isn't tackling an issue you're currently facing but is a safety measure for a potential hazard that could come up in production. You're not facing any random connection losses?

I don't see this influencing the default behavior much, so it looks good to me.

File: avalon/io.py [hunk 1]
```diff
@@ -220,6 +220,24 @@ def decorated(*args, **kwargs):
...
+def auto_reconnect(f):
+    """Handling auto reconnect in 3 retry times"""
+    @functools.wraps(f)
+    def decorated(*args, **kwargs):
+        error = None
+        for retry in range(3):
+            try:
+                return f(*args, **kwargs)
+            except pymongo.errors.AutoReconnect as e:
+                error = e
```
Comment (BigRoy) [inline avalon/io.py]: Do we need this `error = e`, couldn't you just ignore storing the error as variable and just in in the else statement `raise` without a variable to raise the last error that happened? Or am I misinterpreting what this is trying to do?

**Comment by davidlatwe** [comment]
> Though trying three reconnections in this brief time, is that sufficient?

I think the re-connection is actually happening in the next `pymongo` operation, not so sure about the under laying mechanic, but I did simulate the situation by manually block the connection of the primary node and the code worked. :D

> is a safety measure for a potential hazard that could come up in production.  You're not facing any random connection losses?

No, no connection issue here, it's for the safety, yes :)

Thanks @BigRoy !

File: avalon/io.py (hunk 1 above)
Comment (davidlatwe) [reply avalon/io.py]: Oh ! yes, you are right. It's redundant !

**Comment by BigRoy** [comment]
> I think the re-connection is actually happening in the next pymongo operation, not so sure about the under laying mechanic, but I did simulate the situation by manually block the connection of the primary node and the code worked. :D

Did you try the same with a "write" operation? See: https://docs.mongodb.com/manual/replication/#automatic-failover - it states:

> The replica set cannot process write operations until the election completes successfully. The replica set can continue to serve read queries if such queries are configured to run on secondaries while the primary is offline.

So it's likely that read operations are easier to pass, but the write might still fail.
A bit unrelated to the code itself though. :)

**Comment by davidlatwe** [comment]
> Did you try the same with a "write" operation?

Yap, I did try with `insert_one` ;) , and it did wait a bit longer ( must be waiting the election ! ) :P

**Comment by davidlatwe** [comment]
Merging this !!

**Pull request merged**